File: backend/src/main/java/com/autocare/auth/service/AuthService.java
package com.autocare.auth.service;

import com.autocare.auth.dto.CreateUserRequest;
import com.autocare.auth.dto.LoginRequest;
import com.autocare.auth.dto.LoginResponse;
import com.autocare.auth.dto.RefreshTokenRequest;
import com.autocare.auth.dto.RegisterRequest;
import com.autocare.auth.entity.RefreshToken;
import com.autocare.auth.entity.Role;
import com.autocare.auth.entity.User;
import com.autocare.auth.repository.RefreshTokenRepository;
import com.autocare.auth.repository.UserRepository;
import com.autocare.shared.exception.BusinessException;
import com.autocare.shared.exception.ResourceNotFoundException;
import com.autocare.shared.security.JwtService;
import lombok.RequiredArgsConstructor;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class AuthService {

    private static final long REFRESH_TOKEN_VALIDITY_DAYS = 7;

    private final UserRepository userRepository;
    private final RefreshTokenRepository refreshTokenRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final AuthenticationManager authenticationManager;

    @Transactional
    public LoginResponse login(LoginRequest request) {
        // Lança BadCredentialsException em caso de falha, tratada pelo GlobalExceptionHandler (401).
        authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(request.getEmail(), request.getPassword())
        );

        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new ResourceNotFoundException("Usuário não encontrado"));

        String token = jwtService.generateToken(user);
        String refreshToken = issueRefreshToken(user.getId());

        return LoginResponse.builder()
                .token(token)
                .refreshToken(refreshToken)
                .userId(user.getId())
                .name(user.getName())
                .email(user.getEmail())
                .role(user.getRole().name())
                .build();
    }

    /**
     * Troca um refresh token válido por um novo access token (e rotaciona o refresh token,
     * invalidando o anterior). Usado pelo frontend quando o access token expira, evitando
     * obrigar o usuário a logar novamente a cada 24h.
     */
    @Transactional
    public LoginResponse refresh(RefreshTokenRequest request) {
        RefreshToken storedToken = refreshTokenRepository.findByToken(request.getRefreshToken())
                .orElseThrow(() -> new BusinessException("Refresh token inválido"));

        if (storedToken.isRevoked() || storedToken.isExpired()) {
            throw new BusinessException("Refresh token expirado ou revogado. Faça login novamente");
        }

        User user = userRepository.findById(storedToken.getUserId())
                .orElseThrow(() -> new ResourceNotFoundException("Usuário não encontrado"));

        // Rotaciona o refresh token: o antigo não pode mais ser reutilizado.
        refreshTokenRepository.revokeByToken(storedToken.getToken());
        String newRefreshToken = issueRefreshToken(user.getId());
        String newAccessToken = jwtService.generateToken(user);

        return LoginResponse.builder()
                .token(newAccessToken)
                .refreshToken(newRefreshToken)
                .userId(user.getId())
                .name(user.getName())
                .email(user.getEmail())
                .role(user.getRole().name())
                .build();
    }

    @Transactional
    public void logout(RefreshTokenRequest request) {
        refreshTokenRepository.revokeByToken(request.getRefreshToken());
    }

    /**
     * Autocadastro público. Por segurança, SEMPRE cria a conta com o papel mínimo
     * (RECEPTIONIST), independentemente do que for enviado — não existe mais campo
     * "role" no request. Um usuário mal-intencionado não consegue mais se
     * autopromover a ADMIN por aqui (ver CreateUserRequest para criação com papel).
     */
    @Transactional
    public void register(RegisterRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new BusinessException("Este email já está cadastrado");
        }

        User user = User.builder()
                .name(request.getName())
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .role(Role.RECEPTIONIST)
                .isActive(true)
                .build();

        userRepository.save(user);
    }

    /**
     * Criação de usuário com papel arbitrário. Só pode ser chamado por um ADMIN
     * autenticado (a restrição fica no @PreAuthorize do controller).
     */
    @Transactional
    public void createUserByAdmin(CreateUserRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new BusinessException("Este email já está cadastrado");
        }

        Role role;
        try {
            role = Role.valueOf(request.getRole().toUpperCase());
        } catch (IllegalArgumentException e) {
            throw new BusinessException("Papel inválido: " + request.getRole());
        }

        User user = User.builder()
                .name(request.getName())
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .role(role)
                .isActive(true)
                .build();

        userRepository.save(user);
    }

    private String issueRefreshToken(String userId) {
        String token = UUID.randomUUID().toString();
        RefreshToken refreshToken = RefreshToken.builder()
                .token(token)
                .userId(userId)
                .expiresAt(LocalDateTime.now().plusDays(REFRESH_TOKEN_VALIDITY_DAYS))
                .revoked(false)
                .build();
        refreshTokenRepository.save(refreshToken);
        return token;
    }
}

File: backend/src/auth/auth.service.ts
import { Injectable, UnauthorizedException } from "@nestjs/common";
import { JwtService } from "@nestjs/jwt";
import { ConfigService } from "@nestjs/config";
import * as bcrypt from "bcrypt";
import { randomUUID } from "crypto";
import { PrismaService } from "../prisma/prisma.service";
import { BusinessException, ResourceNotFoundException } from "../common/exceptions";
import { LoginDto } from "./dto/login.dto";
import { RegisterDto } from "./dto/register.dto";
import { CreateUserDto } from "./dto/create-user.dto";
import { RefreshTokenDto } from "./dto/refresh-token.dto";
import { Role } from "@prisma/client";

const REFRESH_TOKEN_VALIDITY_DAYS = 7;

@Injectable()
export class AuthService {
  constructor(
    private prisma: PrismaService,
    private jwtService: JwtService,
    private config: ConfigService,
  ) {}

  async login(dto: LoginDto) {
    const user = await this.prisma.user.findUnique({ where: { email: dto.email } });
    // Mensagem genérica de propósito (não revela se o email existe) — equivalente ao
    // BadCredentialsException tratado pelo GlobalExceptionHandler original (401).
    if (!user || !user.isActive || !(await bcrypt.compare(dto.password, user.password))) {
      throw new UnauthorizedException("Email ou senha inválidos");
    }

    const token = this.generateAccessToken(user.id, user.email);
    const refreshToken = await this.issueRefreshToken(user.id);

    return this.buildLoginResponse(user, token, refreshToken);
  }

  async refresh(dto: RefreshTokenDto) {
    const storedToken = await this.prisma.refreshToken.findUnique({
      where: { token: dto.refreshToken },
    });
    if (!storedToken) {
      throw new BusinessException("Refresh token inválido");
    }
    if (storedToken.revoked || storedToken.expiresAt < new Date()) {
      throw new BusinessException("Refresh token expirado ou revogado. Faça login novamente");
    }

    const user = await this.prisma.user.findUnique({ where: { id: storedToken.userId } });
    if (!user) {
      throw new ResourceNotFoundException("Usuário não encontrado");
    }

    // Rotaciona o refresh token: o antigo não pode mais ser reutilizado.
    await this.prisma.refreshToken.update({
      where: { token: storedToken.token },
      data: { revoked: true },
    });
    const newRefreshToken = await this.issueRefreshToken(user.id);
    const newAccessToken = this.generateAccessToken(user.id, user.email);

    return this.buildLoginResponse(user, newAccessToken, newRefreshToken);
  }

  async logout(dto: RefreshTokenDto) {
    await this.prisma.refreshToken.updateMany({
      where: { token: dto.refreshToken },
      data: { revoked: true },
    });
  }

  // Autocadastro público. Por segurança, SEMPRE cria a conta com o papel mínimo
  // (RECEPTIONIST) — não existe campo "role" em RegisterDto. Um usuário
  // mal-intencionado não consegue se autopromover a ADMIN por aqui.
  async register(dto: RegisterDto) {
    const exists = await this.prisma.user.findUnique({ where: { email: dto.email } });
    if (exists) {
      throw new BusinessException("Este email já está cadastrado");
    }

    await this.prisma.user.create({
      data: {
        name: dto.name,
        email: dto.email,
        password: await bcrypt.hash(dto.password, 10),
        role: Role.RECEPTIONIST,
        isActive: true,
      },
    });
  }

  // Criação de usuário com papel arbitrário. Só pode ser chamado por um ADMIN
  // autenticado (a restrição fica no RolesGuard do controller).
  async createUserByAdmin(dto: CreateUserDto) {
    const exists = await this.prisma.user.findUnique({ where: { email: dto.email } });
    if (exists) {
      throw new BusinessException("Este email já está cadastrado");
    }

    const roleValue = dto.role?.toUpperCase();
    if (!Object.values(Role).includes(roleValue as Role)) {
      throw new BusinessException(`Papel inválido: ${dto.role}`);
    }

    await this.prisma.user.create({
      data: {
        name: dto.name,
        email: dto.email,
        password: await bcrypt.hash(dto.password, 10),
        role: roleValue as Role,
        isActive: true,
      },
    });
  }

  private generateAccessToken(userId: string, email: string): string {
    return this.jwtService.sign({ sub: userId, email });
  }

  private async issueRefreshToken(userId: string): Promise<string> {
    const token = randomUUID();
    const expiresAt = new Date();
    expiresAt.setDate(expiresAt.getDate() + REFRESH_TOKEN_VALIDITY_DAYS);

    await this.prisma.refreshToken.create({
      data: { token, userId, expiresAt, revoked: false },
    });
    return token;
  }

  private buildLoginResponse(
    user: { id: string; name: string; email: string; role: Role },
    token: string,
    refreshToken: string,
  ) {
    return {
      token,
      refreshToken,
      userId: user.id,
      name: user.name,
      email: user.email,
      role: user.role,
    };
  }
}
